File: js/storage.js
const Storage = (() => {
  const BASE_URL = '/api';
  function requestSync(endpoint, method = 'GET', body = null) {
    const xhr = new XMLHttpRequest();
    xhr.open(method, BASE_URL + endpoint, false); 
    xhr.setRequestHeader('Content-Type', 'application/json');
    try {
      if (body) {
        xhr.send(JSON.stringify(body));
      } else {
        xhr.send(null);
      }
      if (xhr.status >= 200 && xhr.status < 300) {
        const json = JSON.parse(xhr.responseText);
        return json.ok ? json.data : null;
      }
    } catch (e) {
      console.error("XHR Error:", e);
    }
    return null;
  }
  function getProfiles() {
    return requestSync('/profiles') || [];
  }
  function saveProfile(profile) {
    return requestSync('/profiles', 'POST', profile);
  }
  function deleteProfile(id) {
    return requestSync(`/profiles/${id}`, 'DELETE');
  }
  function getCurrentProfile() {
    const id = localStorage.getItem('parkicare_current_profile');
    if (!id) return null;
    const profiles = getProfiles();
    return profiles.find(p => p.id === id) || null;
  }
  function setCurrentProfile(id) {
    localStorage.setItem('parkicare_current_profile', id);
  }
  function getSessions(profileId, gameType) {
    return requestSync(`/sessions/${profileId}/${gameType}`) || [];
  }
  function saveSession(profileId, gameType, sessionData) {
    return requestSync(`/sessions/${profileId}/${gameType}`, 'POST', sessionData);
  }
  function getWeakProfile(profileId) {
    return requestSync(`/analysis/${profileId}`);
  }
  function runAnalysis(profileId) {
    return requestSync(`/analysis/${profileId}`, 'POST');
  }
  function getProblem(gameType, profileId = null, accessible = true) {
    let url = `/problems/${gameType}?accessible=${accessible}`;
    if (profileId) url += `&profileId=${profileId}`;
    return requestSync(url);
  }
  function login(username, password) {
    const res = requestSync('/auth/login', 'POST', { username, password });
    if (res) {
      localStorage.setItem('parkicare_user', JSON.stringify(res));
      return res;
    }
    return null;
  }
  function register(username, password) {
    const res = requestSync('/auth/register', 'POST', { username, password });
    if (res) {
      localStorage.setItem('parkicare_user', JSON.stringify(res));
      return res;
    }
    return null;
  }
  function logout() {
    requestSync('/auth/logout', 'POST');
    localStorage.removeItem('parkicare_user');
    localStorage.removeItem('parkicare_current_profile');
  }
  function getCurrentUser() {
    const userStr = localStorage.getItem('parkicare_user');
    if (!userStr) return null;
    try {
      return JSON.parse(userStr);
    } catch (e) {
      return null;
    }
  }
  function checkSession() {
    const res = requestSync('/auth/me');
    if (res) {
      localStorage.setItem('parkicare_user', JSON.stringify(res));
      return res;
    } else {
      localStorage.removeItem('parkicare_user');
      return null;
    }
  }
  return {
    getProfiles, saveProfile, deleteProfile,
    getCurrentProfile, setCurrentProfile,
    getSessions, saveSession,
    getWeakProfile, runAnalysis,
    getProblem,
    login, register, logout, getCurrentUser, checkSession
  };
})();

File: py_frontend/storage.py
import json
from pyscript import window

class Storage:
    BASE_URL = '/api'

    @staticmethod
    def request_sync(endpoint, method='GET', body=None):
        try:
            xhr = window.XMLHttpRequest.new()
            xhr.open(method, Storage.BASE_URL + endpoint, False)
            xhr.setRequestHeader('Content-Type', 'application/json')
            if body:
                xhr.send(json.dumps(body))
            else:
                xhr.send(None)
            
            if 200 <= xhr.status < 300:
                resp = json.loads(xhr.responseText)
                return resp.get('data') if resp.get('ok') else None
        except Exception as e:
            print("XHR Error:", e)
        return None

    @staticmethod
    def get_profiles():
        return Storage.request_sync('/profiles') or []

    @staticmethod
    def save_profile(profile):
        return Storage.request_sync('/profiles', 'POST', profile)

    @staticmethod
    def delete_profile(id):
        return Storage.request_sync(f'/profiles/{id}', 'DELETE')

    @staticmethod
    def get_current_profile():
        id = window.localStorage.getItem('parkicare_current_profile')
        if not id:
            return None
        profiles = Storage.get_profiles()
        for p in profiles:
            if p.get('id') == id:
                return p
        return None

    @staticmethod
    def set_current_profile(id):
        window.localStorage.setItem('parkicare_current_profile', id)

    @staticmethod
    def get_sessions(profile_id, game_type):
        all_sessions = Storage.request_sync(f'/sessions/profile/{profile_id}') or []
        return [s for s in all_sessions if s.get('gameType') == game_type]

    @staticmethod
    def save_session(profile_id, game_type, session_data):
        session_data['profileId'] = profile_id
        session_data['gameType'] = game_type
        return Storage.request_sync('/sessions', 'POST', session_data)

    @staticmethod
    def get_weak_profile(profile_id):
        return Storage.request_sync(f'/analysis/{profile_id}')

    @staticmethod
    def run_analysis(profile_id):
        return Storage.request_sync(f'/analysis/{profile_id}')

    @staticmethod
    def get_problem(game_type, profile_id=None, accessible=True):
        url = f'/generate_problem?gameType={game_type}&accessible={"true" if accessible else "false"}'
        if profile_id:
            url += f'&profileId={profile_id}'
        return Storage.request_sync(url)

    @staticmethod
    def login(username, password):
        res = Storage.request_sync('/auth/login', 'POST', {'username': username, 'password': password})
        if res:
            window.localStorage.setItem('parkicare_user', json.dumps(res))
            return res
        return None

    @staticmethod
    def register(username, password):
        res = Storage.request_sync('/auth/register', 'POST', {'username': username, 'password': password})
        if res:
            window.localStorage.setItem('parkicare_user', json.dumps(res))
            return res
        return None

    @staticmethod
    def logout():
        Storage.request_sync('/auth/logout', 'POST')
        window.localStorage.removeItem('parkicare_user')
        window.localStorage.removeItem('parkicare_current_profile')

    @staticmethod
    def get_current_user():
        user_str = window.localStorage.getItem('parkicare_user')
        if not user_str:
            return None
        try:
            return json.loads(user_str)
        except:
            return None

    @staticmethod
    def check_session():
        res = Storage.request_sync('/auth/me')
        if res:
            window.localStorage.setItem('parkicare_user', json.dumps(res))
            return res
        else:
            window.localStorage.removeItem('parkicare_user')
            return None
